Add optimization to avoid load of address

simonvandel opened the pull request:
Look for the sequence
```rust
_2 = &_1;
...
_5 = (*_2)
```

in which we can replace the last statement with `_5 = _1` to avoid the load of _2

rust-highfive commented:
r? @oli-obk

(rust_highfive has picked a reviewer for you, use r? to override)

jonas-schievink commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,49 @@ impl OptimizationFinder<'b, 'tcx> {
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            let statements_to_look_in =
+                self.body.basic_blocks()[location.block].statements[0..stmt_index].iter().rev();
```
This has `O(n²)` performance

simonvandel replied:
True. I originally had it at a lookback of 3, which would be better. It's not guaranteed to optimize all cases. I'll experiment with which heuristic gives the best tradeoff

jonas-schievink commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,49 @@ impl OptimizationFinder<'b, 'tcx> {
+                            _ => {}
+                        }
+                    }
+                    _ => {}
```
What happens in a code sequence like this?

```rust
_1 = &_2;
_1 = _3;
_4 = *_1;
```

If I'm reading the code correctly, this would replace `_4 = *_1` with `_4 = _2`, which is incorrect.

simonvandel replied:
Is it ever possible to have that sequence? I thought SSA form guaranteed only a single assignment to the same variable.

jonas-schievink replied:
MIR is not in SSA form

simonvandel replied:
In that case, yeah, that sequence will get miscompiled. I think it can be fairly easy fixed by bailing if an assignment to _1 is seen that is not of the expected &_1 form. I'll push that up.

bors commented:
:umbrella: The latest upstream changes (presumably #76123) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

simonvandel commented:
Hi @jonas-schievink, thanks for the review.
In the newest commits I:
- Rebased on master to fix conflict
- Added a lookback of 6 (the commit has a table for the numbers I got)
- Added a test and fixed the miscompilation you pointed out

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,58 @@ impl OptimizationFinder<'b, 'tcx> {
+                .rev();
+            for stmt in statements_to_look_in {
+                match &stmt.kind {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r))
```
you also need to bail out on `LlvmAssembly`. Please exhaustively match the `StatementKind`s so we catch all uses.

simonvandel replied:
I added exhaustive matches, and now bail on ASM

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,58 @@ impl OptimizationFinder<'b, 'tcx> {
+                                return Some(());
+                            }
+                            _ => {
+                                // we found an assignment of the place that is not a ref, e.g _local_being_derefed = _2;
```
```suggestion
                                // We found an assignment of the place that is not a ref, e.g `_local_being_derefed = _2;`.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,58 @@ impl OptimizationFinder<'b, 'tcx> {
+                            }
+                            _ => {
+                                // we found an assignment of the place that is not a ref, e.g _local_being_derefed = _2;
+                                // We therefore have to bail the optimization since the optimization is no longer valid
```
```suggestion
                                // We therefore have to bail out of the optimization since the optimization is no longer valid.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,58 @@ impl OptimizationFinder<'b, 'tcx> {
+                            }
+                        }
+                    }
+                    _ => {}
```
if `local_being_derefed` is used on the rhs of an assignment to another local, your optimization may misfire I think.

```rust
    let x = 42;
    let a = 99;
    let mut y = &x;
    let z = &mut y;
    *z = &a;
    println!("{}", *y);
```

will not print `99` after your optimization but `42`

simonvandel replied:
I added a test for this. It does not seem like it will misfire. I also added more clarifying comments on the matches. I have persuaded myself that since we are only applying this optimization on immutable references, we can't have a mutable reference at the same time, so nothing (besides asm) can break this optimization. My reasoning may be wrong though..

simonvandel replied:
You are correct. Setting the lookback higher does indeed cause a miscompilation. I'll look into fixing it.

oli-obk replied:
The general way to look for mutation is to define a visitor and visit the construct you want to analyze (`Statement` in your case) and implement https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/trait.Visitor.html#method.visit_local to and check the `PlaceContext` if the local matches the local you want to check.

simonvandel replied:
Ah, I saw your comment a tad too late. I already pushed a fix, but i'll see if it is cleaner/less adhoc to use the visitor. Thanks for the hint.

simonvandel replied:
The latest commits now check that local_being_derefed is not mutated using a visitor. I have verified that setting the lookback to 10000 causes a misoptimization without the fix, and fixes it with the fix. I have set it back to to avoid bad runtime complexity. I'm not sure how best to represent that the fix is actually working. For now I separated it into two commits.

oli-obk replied:
Thanks! Splitting it into individual commits made this great to review.

I also don't know how to test it. A bogus idea (Don't do this :laughing:): instead of `6` use `6 + mir_opt_level`, then we can do `-Zmir-opt-level=99999` on that test.

An alternative to having this magic number would be to implement the optimization as a feed forward (am I using words correctly here?) optimization. What I mean is that we don't go through every deref and look back, but we walk forward through each block and keep a set of "deref-optimizable" locals that we update as we go.

oli-obk replied:
That said, I'm fine with merging this optimization as long as there's a tracking issue for exploring such a change to the optimization. I'm not sure whether that change is feasible in practice, but we should explore it

simonvandel replied:
Cool. Yeah it should be possible to do this optimization  without the lookback, such that a single statement is only visited once. I'll open a new pr for that.

Can we do a perf run on the current pr? I'm curious if this has any impact

oli-obk reviewed (changes requested):
This optimization is very scary, it has lots of edge cases that we need to think about

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            // 6 is chosen as a heuristic determined by seeing the number of times the optimization kicked in compiling rust std
+            let lower_index = stmt_index.saturating_sub(6);
```
I believe the only reason my example does not get misoptimized is this magic number. That is too fragile for my liking and this fact is also not documented.

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+            }?;
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
```
```suggestion
            // Look behind for statement that assigns the local from a address of operator.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+
+            let stmt_index = location.statement_index;
+            // look behind for statement that assigns the local from a address of operator
+            // 6 is chosen as a heuristic determined by seeing the number of times the optimization kicked in compiling rust std
```
```suggestion
            // 6 is chosen as a heuristic determined by seeing the number of times
            // the optimization kicked in compiling rust std.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
     }
 
+    fn find_deref_of_address(&mut self, rvalue: &Rvalue<'tcx>, location: Location) -> Option<()> {
+        // look for the sequence
```
```suggestion
        // Look for the sequence
        //
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+        // ...
+        // _5 = (*_2);
+        //
+        // which we can replace the last statement with `_5 = _1;` to avoid the load of _2
```
```suggestion
        // which we can replace the last statement with `_5 = _1;` to avoid the load of `_2`.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+            for stmt in statements_to_look_in {
+                match &stmt.kind {
+                    // Exhaustive match on statements to detect conditions that warrant we bail out of the optimization.
+                    // by only looking applying the optimization for immutable references e.g _local_being_deref = &_1;
+                    // we only need to check that opaque statements
+                    // like inline asm are not in the statements which can still modify `local_being_derefed`
+                    // or the value of `*local_being_derefed`
```
```suggestion
                    // By only looking applying the optimization for immutable references e.g _local_being_deref = &_1;
                    // we only need to check that opaque statements
                    // like inline asm are not in the statements which can still modify `local_being_derefed`
                    // or the value of `*local_being_derefed`.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+                    rustc_middle::mir::StatementKind::Assign(box (l, r)) => {
+                        if l.local == local_being_derefed {
+                            match r {
+                                // looking for immutable reference e.g _local_being_deref = &_1;
```
```suggestion
                                // Looking for immutable reference e.g `_local_being_deref = &_1;`.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+                                _ => return None,
+                            }
+                        }
+                        // at this point we know that `local_being_derefed` is not being assigned to.
```
```suggestion
                        // At this point we know that `local_being_derefed` is not being assigned to.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+                            }
+                        }
+                        // at this point we know that `local_being_derefed` is not being assigned to.
+                        // Since there is an immutable borrow on `local_being_derefed`, it is not possible to take a mutable reference to `local_being_derefed`
```
```suggestion
                        // Since there is an immutable borrow on `local_being_derefed`,
                        // it is not possible to take a mutable reference to `local_being_derefed`
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+                        // which would be required to modify it.
+                    }
+
+                    // These can not affect/mutate local_being_derefed, so ignore these
```
```suggestion
                    // These can not affect/mutate local_being_derefed, so ignore these.
```

oli-obk commented on compiler/rustc_mir/src/transform/instcombine.rs:
```diff
@@ -87,6 +93,81 @@ impl OptimizationFinder<'b, 'tcx> {
+                    | rustc_middle::mir::StatementKind::AscribeUserType(_, _)
+                    | rustc_middle::mir::StatementKind::SetDiscriminant { .. } => {}
+
+                    // Inline asm can do anything so bail out of the optimization
```
```suggestion
                    // Inline asm can do anything, so bail out of the optimization.
```

oli-obk commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 116283b910a7f8809260ba91cb9b9b647a8900a3 with merge 7d835e12834e0230779883b269523b01a940c2df...
<!-- homu: {"type":"TryBuildStarted","head_sha":"116283b910a7f8809260ba91cb9b9b647a8900a3","merge_sha":"7d835e12834e0230779883b269523b01a940c2df"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1139955865), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36726)
Build commit: 7d835e12834e0230779883b269523b01a940c2df (`7d835e12834e0230779883b269523b01a940c2df`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36726","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1139955865"},"merge_sha":"7d835e12834e0230779883b269523b01a940c2df"} -->

rust-timer commented:
Queued 7d835e12834e0230779883b269523b01a940c2df with parent 10b3595ba6a4c658c9dea105488fc562c815e434, future [comparison URL](https://perf.rust-lang.org/compare.html?start=10b3595ba6a4c658c9dea105488fc562c815e434&end=7d835e12834e0230779883b269523b01a940c2df).

rust-timer commented:
Finished benchmarking try commit (7d835e12834e0230779883b269523b01a940c2df): [comparison url](https://perf.rust-lang.org/compare.html?start=10b3595ba6a4c658c9dea105488fc562c815e434&end=7d835e12834e0230779883b269523b01a940c2df).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

oli-obk commented:
perf looks ok. Please squash so we get all the back and forth out of the timeline

simonvandel commented:
Squashed and rebased on master

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 4dedb76438b94788182133633558ebd8382115fb has been approved by `oli-obk`

<!-- @bors r=oli-obk 4dedb76438b94788182133633558ebd8382115fb -->
<!-- homu: {"type":"Approved","sha":"4dedb76438b94788182133633558ebd8382115fb","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 4dedb76438b94788182133633558ebd8382115fb with merge 40fe3db7186781f895de3f6d860dbd5d570ec21c...
<!-- homu: {"type":"BuildStarted","head_sha":"4dedb76438b94788182133633558ebd8382115fb","merge_sha":"40fe3db7186781f895de3f6d860dbd5d570ec21c"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1141670242)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1141670242","builder_name":"checks-actions"} -->

oli-obk commented:
Looks like you need to rebase and rebless again?

simonvandel commented:
Rebased on master and added a new commit such that Add is always generated instead of CheckedAdd, which broke the test diff on noopt testruns.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit dfc469ddae5b13a611f70eef27e6d60dbbce3445 has been approved by `oli-obk`

<!-- @bors r=oli-obk dfc469ddae5b13a611f70eef27e6d60dbbce3445 -->
<!-- homu: {"type":"Approved","sha":"dfc469ddae5b13a611f70eef27e6d60dbbce3445","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit dfc469ddae5b13a611f70eef27e6d60dbbce3445 with merge f47df31ae5d7d9795399dca3a0003c1856900361...
<!-- homu: {"type":"BuildStarted","head_sha":"dfc469ddae5b13a611f70eef27e6d60dbbce3445","merge_sha":"f47df31ae5d7d9795399dca3a0003c1856900361"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1146935791), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36760)
Approved by: oli-obk
Pushing f47df31ae5d7d9795399dca3a0003c1856900361 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36760","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1146935791"},"merge_sha":"f47df31ae5d7d9795399dca3a0003c1856900361"} -->

eddyb commented:
@rust-lang/wg-mir-opt I think this would be better suited to constant-folding with a notion of "symbolic" values.
(not necessarily symbolic miri evaluation, but propagatable abstract values that represent runtime values which are effectively "constant" within one execution of the function)

But also I'm a bit worried this kind of optimization may break stacked borrows type assumptions if the indirection was "weakening" the semantics of the access, and making it direct "strengthens" it too much - though this may only be a problem with raw pointers.
